numismatic/feeds/base.py

```diff
@@ -241,18 +242,39 @@ def _make_request(self, api_url, params=None, headers=None, raw=False):
             data = response
         return data
 
-
 @attr.s
 class WebsocketClient(abc.ABC):
     '''Base class for WebsocketClient feeds'''
     # FIXME: Is this really an ABC? What abstractmethods are there?
 
-    exchange = None
-    websocket_url = None
-    websocket = attr.ib(default=None)
+    exchange = attr.ib(default=None)
+    websocket_url = attr.ib(default=None)
+    websocket = attr.ib(default=None, repr=False)
+    subscriptions = attr.ib(default=attr.Factory(list), repr=False)
+
+    def __attrs_post_init__(self):
+        if self.exchange is None:
+            self.exchange = self.__class__.exchange
+        if self.websocket_url is None:
+            self.websocket_url = self.__class__.websocket_url
+        asyncio.ensure_future(self._connect())
+        asyncio.ensure_future(self._listener())
+
+    async def _connect(self):
+        '''
+            Connects to websocket. Uses a future to ensure that only one
+            connection at a time will happen
+        '''
+        if self.websocket is None:
+            logger.info(f'Connecting to {self.websocket_url!r} ...')
+            self.websocket = \
+                asyncio.ensure_future(websockets.connect(self.websocket_url))
+        if isinstance(self.websocket, asyncio.Future):
+            self.websocket = await self.websocket
+        return self.websocket
```

Comment:
You're right, it shouldn't return the websocket anymore. I was trying to avoid having a method with a side-effect as I don't like them and try to avoid them wherever possible. One of my intermediate commits did achieve this and that code was much neater in my opinion. However it sacrificed the concurrency in connecting to the websocket. I think that's an acceptable cost and won't make any difference in production use cases but it made startup slower in interactive sessions in the shell so I opted against it for now. We could always bring it back again later.